logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/auth/LoginUseCase.kt

```diff
@@ -42,16 +43,21 @@ class LoginUseCase(
             }
             else -> return@suspending AuthenticationResult.Failure.InvalidUserIdentifier
         }.coFold({
-            when (it.kaliumException) {
-                is KaliumException.InvalidRequestError -> {
-                    if (it.kaliumException.isInvalidCredentials()) {
-                        AuthenticationResult.Failure.InvalidCredentials
-                    } else {
-                        AuthenticationResult.Failure.Generic(it)
-                    }
-                }
-                else -> AuthenticationResult.Failure.Generic(it)
-            }
+                 when(it) {
+                     is NetworkFailure.ServerMiscommunication -> {
+                         when (it.kaliumException) {
+                             is KaliumException.InvalidRequestError -> {
+                                 if (it.kaliumException.isInvalidCredentials()) {
+                                     AuthenticationResult.Failure.InvalidCredentials
+                                 } else {
+                                     AuthenticationResult.Failure.Generic(it)
+                                 }
+                             }
+                             else -> AuthenticationResult.Failure.Generic(it)
+                         }
```

Comment:
A bit too deep for quick understanding. What about extracting this to another function?